### DISCUSSION TITLE
spark 101

### DISCUSSION THREAD

**Discussion by TinyHust** [TinyHust/notes#431] [DE]
**Examples:**
- https://github.com/mediumhust/tokyo-olympic-azure-data-engineering-project
- https://github.com/sudarshan-koirala/30-days-of-Databricks

**Comment by mediumhust** [comment]
[dataframe style](https://www.youtube.com/watch?v=38k6duccvd4&list=PLLr_PBm86prvN72CeX9aKqiE3zLValzKC&index=22)
- programmatic
- column expression
- sql

### terms:
- skew partition
- [Cluster Manager Types](https://spark.apache.org/docs/latest/cluster-overview.html#cluster-manager-types)
  - [Standalone](https://spark.apache.org/docs/latest/spark-standalone.html)
  - [Apache Mesos](https://spark.apache.org/docs/latest/running-on-mesos.html)
  - [Hadoop YARN](https://spark.apache.org/docs/latest/running-on-yarn.html)
  - [Kubernetes](https://spark.apache.org/docs/latest/running-on-kubernetes.html)
- deployment modes:
  - client mode
  - cluster mode
- write 
  - modes:
    - append
    - overwrite
    - error or errorifexists
    - ignore
  - options:
    - mergeSchema (schema evolution)
- read 
  - modes:
    - permissive (default)
    - dropmalformed
    - failfast
  - options:
    - inferSchema (recommended false in prod)
    - samplingRatio
    - dateFormat
- repartition vs coalesce
- cache
- DAG (Directed Acyclic Graph)
- AQE (Adaptive Query Execution)
- joins:
  - Broadcast Hash Join
  - Shuffle Sort Merge Join
  - Shuffle Hash Join 
-  partitionBy vs bucketBy
- logical plan `.explain(True)`
  - Parsed
  - Analyzed
    - Catalyst Optimizer 
  - Optimized
- physical plan
- flow: `spark sql/dataframe` > `logical plan` > `physical plan` > `cost model` >` selected physical plan` > `rdd`
- metastore
  - https://medium.com/@sarfarazhussain211/metastore-in-apache-spark-9286097180a4
  - https://rafaelrampineli.medium.com/understanding-the-role-of-hive-meta-store-in-spark-sql-and-how-databricks-organizes-databases-41b1747ea746

**Comment by TinyHust** [comment]
pyspark write partition by

`sample.csv`
```
201911240130 a
201911250132 b
201911250143 c
201911250223 z
201911250224 d
```

```python
custom_header = "timestamp\tvalue"
schema = StructType()
col_names = custom_header.split("\t")
for c in col_names:
    schema.add(StructField(c.strip(), StringType()))

df = spark.read.csv("hdfs://sample.csv", header=False, sep="\t", schema=schema)

df_final = df.withColumn("timestamp", to_timestamp(col("timestamp"), 'yyyyMMddHHmm')) \
           .withColumn("year", date_format(col("timestamp"), "yyyy")) \
           .withColumn("month", date_format(col("timestamp"), "MM")) \
           .withColumn("day", date_format(col("timestamp"), "dd")) \
           .withColumn("hour", date_format(col("timestamp"), "HH")) \
           .drop("timestamp")

df_final.show(truncate=False)
```
```
+-----+----+-----+---+----+
|value|year|month|day|hour|
+-----+----+-----+---+----+
|a    |2019|11   |24 |01  |
|b    |2019|11   |25 |01  |
|c    |2019|11   |25 |01  |
|z    |2019|11   |25 |02  |
|d    |2019|11   |25 |02  |
+-----+----+-----+---+----+
```
```python
df_final.write.partitionBy("year", "month", "day", "hour") \
    .mode("overwrite") \
    .option("header", "false").option("sep", "\t") \
    .csv("hdfs://dest/")
```

```
hdfs://dest/year=2019/month=11/day=24/hour=01/part-0000-xxx.csv
```

**Comment by mediumhust** [comment]
### memory
```python
spark.conf.set("spark.executor.memory", "2g") # Heap memory
spark.conf.set("spark.memory.fraction", "0.6") # User memory
spark.conf.set("spark.memory.storageFraction", "0.5") # Storage memory
# Off-heap memory = Max(384Mb, 10% of spark.executor.memory)
```
